File: Testing-001/solution/Solution.MyApplication/PersonService.cs
using Solution.MyApplication.Validation;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Solution.MyApplication
{
    public class PersonService
    {
        /// <summary>
        /// Gets or sets the person validator.
        /// </summary>
        /// <value>
        /// The person validator.
        /// </value>
        public IValidation PersonValidator { get; set; }

        public Func<DateTime> GetTodaysDate { get; set; }

        /// <summary>
        /// Initializes a new instance of the <see cref="PersonService"/> class.
        /// </summary>
        /// <param name="personValidator">The person validator.</param>
        /// <param name="retrieveCurrentDate">
        /// Returns the current date. NB: Not supported by Unity, but can be done with Castle. 
        /// This functionality could be contained within a DateManager class (or use Noda time!)
        /// </param>
        public PersonService(IValidation personValidator, Func<DateTime> retrieveCurrentDate )
        {
            PersonValidator = personValidator;
            GetTodaysDate = retrieveCurrentDate;

            // Ideally this would be a DB call
            AgeGroupMap = new Dictionary<int,AgeGroup>();

            AgeGroupMap[12] = AgeGroup.Child;
            AgeGroupMap[17] = AgeGroup.Teen;
            AgeGroupMap[25] = AgeGroup.YoungAdult;
            AgeGroupMap[74] = AgeGroup.Adult;
            AgeGroupMap[999] = AgeGroup.Retired;
        }

        /// <summary>
        /// Gets or sets the age group map.
        /// </summary>
        /// <remarks>
        /// The map = 
        ///     key: max age in bracket
        ///     value: corresponding Age Group for bucket
        /// </remarks>
        /// <value>
        /// The age group map.
        /// </value>
        public Dictionary<int, AgeGroup> AgeGroupMap { get; set; }

        /// <summary>
        /// Gets the age group for a particular customer.
        /// </summary>
        /// <param name="customer">The customer to calculate the AgeGroup of.</param>
        /// <returns>The correct age group for the customer</returns>
        /// <exception cref="System.ApplicationException">If the customer is invalid</exception>
        public AgeGroup GetAgeGroup(Person customer)
        {
            if (!PersonValidator.IsValid(customer))
            {
                throw new ApplicationException("customer is invalid");
            }

            // Calculate age
            int age = CalculateAge(customer.Dob);

            // Return the correct age group
            var viableBuckets = AgeGroupMap.Where(x => x.Key >= age);
            return AgeGroupMap[viableBuckets.Min(x => x.Key)];
        }

        /// <summary>
        /// Calculates age.
        /// </summary>
        /// <param name="dateOfBirth">The date of birth.</param>
        /// <returns>The age (in whole years)</returns>
        public int CalculateAge(DateTime dateOfBirth)
        {
            // NB: should really be in a utility class, but it's here for the purpose of the workshop!
            DateTime zeroTime = new DateTime(1, 1, 1);
            int age = (zeroTime + (GetTodaysDate().Date - dateOfBirth)).Year - 1;

            return age;
        }
    }
}

File: Testing-001-Java/src/main/java/com/tomkuhn/unnittest/solution/PersonService.java
package com.tomkuhn.unnittest.solution;

import com.tomkuhn.unnittest.solution.validation.Validator;

import java.time.LocalDate;
import java.time.Period;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

public class PersonService {
    private final Validator personValidator;
    private Map<Integer, AgeGroup> ageGroupMap = new HashMap<>();

    private Supplier<LocalDate> getLocalDateNow;

    public PersonService(Validator personValidator, Supplier<LocalDate> getNow) {

        this.personValidator = personValidator;
        this.getLocalDateNow = getNow;

        ageGroupMap.put(12, AgeGroup.CHILD);
        ageGroupMap.put(17, AgeGroup.TEEN);
        ageGroupMap.put(25, AgeGroup.YOUNG_ADULT);
        ageGroupMap.put(74, AgeGroup.ADULT);
        ageGroupMap.put(999, AgeGroup.RETIRED);
    }

    public Map<Integer, AgeGroup> getAgeGroupMap() {
        return ageGroupMap;
    }

    public void setAgeGroupMap(Map<Integer, AgeGroup> ageGroupMap) {
        this.ageGroupMap = ageGroupMap;
    }

    public AgeGroup getAgeGroup(Person customer) {
        if (!personValidator.isValid(customer)) {
            throw new IllegalArgumentException("The customer is invalid");
        }

        int age = calculateAge(customer.getDob());

        int smallestViableAgeGroup = ageGroupMap.entrySet().stream()
                                        .mapToInt(Map.Entry::getKey)
                                        .filter(x -> x >= age)
                                        .min().orElse(999);

        return ageGroupMap.get(smallestViableAgeGroup);
    }

    public int calculateAge(LocalDate dob) {
        return Period.between(dob, getLocalDateNow.get()).getYears();
    }
}
